### PR TITLE
Add x-www-form-urlencoded content type for ApiRest request

### LINKED ISSUE #92: Content Type x-www-form-urlencoded allowed for Api REST
Summary

Content type x-www-form-urlencoded for API Rest calls are not allowed as Magento does not have any deserializer.

Proposed solution

Add new deserializer class for this type of request, and include new type in deserializers argument of Magento\Framework\Webapi\Rest\Request\DeserializerFactory.
  

Code content in https://github.com/magento/magento2/pull/18463

### PR DESCRIPTION
**Pull request by i4rebecamartinez**
### Description

Content type x-www-form-urlencoded for API Rest calls are not allowed as Magento does not have any deserializer.

Fixes https://github.com/magento/community-features/issues/92

### Proposed solution

Add new deserializer class for this type of request, and include new type in deserializers argument of Magento\Framework\Webapi\Rest\Request\DeserializerFactory.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18463) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=18463) before we can accept your contribution.<br/>

**Comment by misha-kotov** [comment]
@i4rebecamartinez can you elaborate on the use cases and the impact of this issue? it's not very clear from the description.

**Comment by sidolov** [comment]
Hi @i4rebecamartinez , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by paulpartington-cti** [comment]
I have recently come across a need for this were we provided an API endpoint for a payment gateway to notify us of order statuses. Unfortunately I had no control of the headers they sent in their updates and the API would throw a 500 error when the header was `application/x-www-form-urlencoded`.

I brought this change into my project and it worked.

I'd like to re-open this PR as it would be a useful change to incorporate.

**Comment by novikor** [comment]
Hi, @i4rebecamartinez ! Thank you for your contribution.
Please sign the Adobe CLA

File: lib/internal/Magento/Framework/Webapi/Rest/Request/Deserializer/FormUrlencoded.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+namespace Magento\Framework\Webapi\Rest\Request\Deserializer;
+
+use Magento\Framework\App\State;
+use Magento\Framework\Phrase;
+use InvalidArgumentException;
+use Magento\Framework\Webapi\Rest\Request\DeserializerInterface;
+use Zend\Stdlib\Parameters;
+
+/**
+ * Class FormUrlencoded
```
Comment (novikor) [inline lib/internal/Magento/Framework/Webapi/Rest/Request/Deserializer/FormUrlencoded.php]: It seems that the class PHPDoc could be removed or replaced by its description

File: lib/internal/Magento/Framework/Webapi/Rest/Request/Deserializer/FormUrlencoded.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * FormUrlencoded deserializer of REST request content.
+ *
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
```
Comment (novikor) [inline lib/internal/Magento/Framework/Webapi/Rest/Request/Deserializer/FormUrlencoded.php:7]: Please use strict type mode [(1.3.1)](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#1-basic-programming-principles)

**Review by novikor — CHANGES_REQUESTED** [review]
Good job, @i4rebecamartinez !
Please check my comments.

It seems that the changes must be covered with [WebApi](Magento\Framework\Webapi\Rest\Request\DeserializerInterface) tests.

What do you think about implementing the test coverage and updating the description with more specific steps to reproduce?

**Comment by ihor-sviziev** [comment]
Please also use \Laminas\ instead of \Zend\

**Comment by ihor-sviziev** [comment]
Hi @i4rebecamartinez @paulpartington-cti,

In case if you need to get some requests from external services you have two options:
1. implement API endpoint and client should correctly use it
2. implement controller that will handle requests from external system, and you could put any logic in here
3. you could apply such change in your custom extension, if you really need it

Looks like you tried to do 1st option, but client wasn't correctly using it. So in this case you need to switch to 2nd option.

Personally I don't think such change should be merged to Magento and also this PR and original issue were created in 2018 and didn't get any attention from community - so it means it's not frequent issue that will be helpful for many people, but will definitely additional cases -> will increase complexity and potentially new issues could appear because of that.

Thank you for understanding. If you'll have any other use cases - please feel free to describe them.

**Pull request closed without merging**